File: 179 Largest Number/Solution.java
import java.util.Arrays;
import java.util.Comparator;

class Solution {
    public static String largestNumber(int[] nums) {
        // 1. Convert int array to String array so we can sort properly
        String[] strNums = new String[nums.length];
        for (int i = 0; i < nums.length; i++) {
            strNums[i] = String.valueOf(nums[i]);
        }

        // 2. Custom Comparator
        // If s1 + s2 > s2 + s1, then s1 should come first.
        Arrays.sort(strNums, new Comparator<String>() {
            @Override
            public int compare(String s1, String s2) {
                String order1 = s1 + s2;
                String order2 = s2 + s1;
                // We want descending order, so we compare order2 against order1
                return order2.compareTo(order1);
            }
        });

        // 3. Edge Case: If the largest number is "0", the result is just "0"
        // (e.g. input [0, 0] should return "0", not "00")
        if (strNums[0].equals("0")) {
            return "0";
        }

        // 4. Build the result
        StringBuilder sb = new StringBuilder();
        for (String s : strNums) {
            sb.append(s);
        }
        return sb.toString();
    }
    public static void main(String[] args) {
        System.out.println(largestNumber(new int[]{10, 2}));                                    //"210"
        System.out.println(largestNumber(new int[]{3,30,34,5,9}));                              //"9534330"
    }
}

File: 179 Largest Number/Solution.py
from functools import cmp_to_key

class Solution:
    def largestNumber(self, nums: list[int]) -> str:
        # 1. Convert all integers to strings
        # We need strings to perform concatenation checks (e.g., "9" + "30")
        str_nums = list(map(str, nums))
        
        # 2. Define the Custom Comparator
        # Returns 1 if x should come AFTER y
        # Returns -1 if x should come BEFORE y
        # Returns 0 if they are equal
        def compare(x, y):
            if x + y > y + x:
                return 1  # x is "larger"
            elif x + y < y + x:
                return -1 # x is "smaller"
            else:
                return 0
        
        # 3. Sort using the comparator
        # We use reverse=True because we want the "largest" combinations first
        str_nums.sort(key=cmp_to_key(compare), reverse=True)
        
        # 4. Join the list into a single string
        largest_num = "".join(str_nums)
        
        # 5. Edge Case: Handle [0, 0] -> "00"
        # If the result starts with '0', the number is just 0
        if largest_num[0] == "0":
            return "0"
            
        return largest_num

if __name__ == "__main__":
    sol = Solution()
    print(sol.largestNumber([3, 30, 34, 5, 9])) # Output: "9534330"